### PR TITLE
`deploy_at`

### PR DESCRIPTION
**Pull request by piotmag769**
## Introduced changes

- `deploy_at` added to `snforge_std`

## Checklist

- [x] Linked relevant issue
- [x] Updated relevant documentation
- [x] Added relevant tests
- [x] Performed self-review of the code
- [x] Added changes to `CHANGELOG.md`

### CODE REVIEW

File: crates/forge/src/cheatcodes_hint_processor.rs [hunk 1]
```diff
@@ -339,25 +338,29 @@ impl CairoHintProcessor<'_> {
...
-                    Err(CheatcodeError::Unrecoverable(err)) => Err(err),
-                }
+                let calldata_length = inputs[1].to_usize().unwrap();
+                let calldata = Vec::from(&inputs[2..(2 + calldata_length)]);
+
+                let contract_address_felt = inputs[2 + calldata_length].clone();
+                let contract_address = ContractAddress(
+                    PatriciaKey::try_from(felt_to_stark_felt(&contract_address_felt))
+                        .expect("StarkFelt to PatriciaKey conversion failed"),
+                );
```
Comment (war-in) [inline crates/forge/src/cheatcodes_hint_processor.rs:357]: `contract_class_from_felt` is here too

File: crates/forge/src/cheatcodes_hint_processor.rs [hunk 2]
```diff
@@ -489,6 +492,30 @@ impl CairoHintProcessor<'_> {
     }
 }
 
+fn handle_deploy_result(
+    deploy_result: Result<ContractAddress, CheatcodeError>,
+    buffer: &mut MemBuffer,
+) -> Result<(), EnhancedHintError> {
+    match deploy_result {
+        Ok(contract_address) => {
+            let felt_contract_address: Felt252 = stark_felt_to_felt(*contract_address.0.key());
```
Comment (war-in) [inline crates/forge/src/cheatcodes_hint_processor.rs:501]: `contract_address_to_felt` 😉

File: crates/forge/tests/data/contracts/deploy_at_checker.cairo [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+}
+
+#[starknet::contract]
+mod DeployAtChecked {
+    use starknet::ContractAddress;
+    use traits::Into;
+
+    #[external(v0)]
+    fn get_caller_address(ref self: ContractState, contract_address: ContractAddress) -> felt252 {
+        IPrankChecker{ contract_address }.get_caller_address
```
Comment (war-in) [inline crates/forge/tests/data/contracts/deploy_at_checker.cairo:13]: Shouldn't it be called? `()`

File: crates/forge/tests/integration/deploy_at.rs [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        use array::ArrayTrait;
+        use starknet::ContractAddress;
+        
+        #[starknet::interface]
+        trait IProxy<TContractState> {
+            fn get_caller_address(ref self: TContractState, checker_address: ContractAddress) -> felt252;
+        }
+
+        #[test]
+        fn test_deploy_error_handling() {
```
Comment (war-in) [inline crates/forge/tests/integration/deploy_at.rs:22]: Let's change the name

File: crates/forge/tests/data/contracts/deploy_at_checker.cairo (hunk 3 above)
Comment (war-in) [reply crates/forge/tests/data/contracts/deploy_at_checker.cairo:13]: Actually, I think it is just a leftover. We should remove it

File: docs/src/appendix/forge-library/deploy_at.md [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+#[test]
+fn test_deploy() {
+    let contract = declare('HelloStarknet');
+
+    let mut constructor_calldata = ArrayTrait::new();
+    constructor_calldata.append(42_u8.into());
+    constructor_calldata.append(21);
+    constructor_calldata.append(37);
+    
+    let contract_address = 123.try_into().unwrap();
```
Comment (war-in) [inline docs/src/appendix/forge-library/deploy_at.md:37]: How about using `contract_address_const`?

File: docs/src/appendix/forge-library/deploy_at.md [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    fn deploy_at(
+        self: @ContractClass,
+        constructor_calldata: @Array::<felt252>,
+        contract_address: ContractAddress
+    ) -> Result<ContractAddress, RevertedTransaction>;
+    // ...
+}
+```
+
+Deploys a contract at given address and returns the address.
```
Comment (Radinyn) [inline docs/src/appendix/forge-library/deploy_at.md:14]:
```suggestion
Deploys a contract at a given address and returns the address.
```

File: snforge_std/src/cheatcodes/contract_class.cairo [hunk 7]
```diff
@@ -31,6 +31,11 @@ trait ContractClassTrait {
     fn deploy(
         self: @ContractClass, constructor_calldata: @Array::<felt252>
     ) -> Result<ContractAddress, RevertedTransaction>;
+    fn deploy_at(
```
Comment (Radinyn) [inline snforge_std/src/cheatcodes/contract_class.cairo:34]:
```suggestion
    ) -> Result<ContractAddress, RevertedTransaction>;
    
    fn deploy_at(
```

File: crates/forge/tests/integration/deploy_at.rs
Comment (war-in) [inline crates/forge/tests/integration/deploy_at.rs:1]: I wonder if we need more tests here 🤔 We do have most of them on the cheatnet side but maybe checking an error message wouldn't be such a bad idea. Wdyt?

**Review by war-in — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 8]
```diff
@@ -11,6 +11,16 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 #### Added
 
+- `deploy_at` function
```
Comment (Radinyn) [inline CHANGELOG.md:14]:
```suggestion
- `deploy_at` cheatcode
```

is this a cheatcode?

**Review by Radinyn — APPROVED** [review]

File: crates/cheatnet/src/cheatcodes/deploy.rs [hunk 9]
```diff
@@ -23,9 +23,11 @@ impl CheatnetState {
         &mut self,
         class_hash: &ClassHash,
         calldata: &[Felt252],
-        salt: ContractAddressSalt,
         contract_address: ContractAddress,
     ) -> Result<ContractAddress, CheatcodeError> {
+        let salt = self.get_salt();
+        self.increment_deploy_salt_base();
```
Comment (cptartur) [inline crates/cheatnet/src/cheatcodes/deploy.rs:29]: I'm not sure why was this moved here. Won't it influence the contract address?

File: crates/forge/tests/data/contracts/deploy_at_checker.cairo
[the remark this answers is not in the document]
Comment (cptartur) [reply crates/forge/tests/data/contracts/deploy_at_checker.cairo:1]: Because it doesn't seem like it

File: crates/forge/tests/integration/deploy_at.rs
Comment (cptartur) [reply crates/forge/tests/integration/deploy_at.rs:1]: Though perhaps some test for unsuccessful case could be useful

File: docs/src/appendix/forge-library/deploy_at.md [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+
+    let mut constructor_calldata = ArrayTrait::new();
+    constructor_calldata.append(42_u8.into());
+    constructor_calldata.append(21);
+    constructor_calldata.append(37);
+    
+    let contract_address = 123.try_into().unwrap();
+
+    let address = contract.deploy_at(@constructor_calldata, contract_address).unwrap();
+    // ...
```
Comment (cptartur) [inline docs/src/appendix/forge-library/deploy_at.md:40]: Perhaps adding an assert for `address` here would make it extra explicit what this function does?

File: crates/cheatnet/src/cheatcodes/deploy.rs (hunk 9 above)
Comment (war-in) [reply crates/cheatnet/src/cheatcodes/deploy.rs:29]: I think it is ok to move it here. Previously `deploy_at` didn't change the salt so I think deploying two same contracts with the `deploy_at` function would panic. 
@piotmag769 could you please add an issue (or a test) for checking it?

File: CHANGELOG.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (piotmag769) [reply CHANGELOG.md:14]: Hmmm, is it a cheatcode though? I guess you are right

File: crates/forge/src/cheatcodes_hint_processor.rs [hunk 11]
```diff
@@ -339,25 +338,29 @@ impl CairoHintProcessor<'_> {
...
-                        let felt_contract_address: Felt252 =
-                            stark_felt_to_felt(*contract_address.0.key());
+                handle_deploy_result(
+                    self.cheatnet_state.deploy(&class_hash, &calldata),
+                    &mut buffer,
+                )
+            }
+            "deploy_at" => {
+                let class_hash = inputs[0].clone();
+                let class_hash = ClassHash(StarkFelt::new(class_hash.to_be_bytes()).unwrap());
```
[the remark this answers is not in the document]
Comment (piotmag769) [reply crates/forge/src/cheatcodes_hint_processor.rs:348]: Yup but in cheatnet, I cannot create cyclic dependency

File: crates/forge/src/cheatcodes_hint_processor.rs (hunk 1 above)
Comment (piotmag769) [reply crates/forge/src/cheatcodes_hint_processor.rs:357]: Same as above

File: crates/cheatnet/src/cheatcodes/deploy.rs (hunk 9 above)
Comment (piotmag769) [reply crates/cheatnet/src/cheatcodes/deploy.rs:29]: It is to avoid having to pass salt to `deploy_at` so it can be used directly from `forge` as salt is stored in cheatnet

I think we already have a test for that, don't we? `deploy_two_at_the_same_address` in cheatnet

File: crates/forge/tests/integration/deploy_at.rs
Comment (piotmag769) [reply crates/forge/tests/integration/deploy_at.rs:1]: I can add a test with deploying 2 contracts at the same address, WDYT?

File: docs/src/appendix/forge-library/deploy_at.md (hunk 5 above)
Comment (piotmag769) [reply docs/src/appendix/forge-library/deploy_at.md:37]: We use the pattern I used everywhere in docs so I would leave it this way

**Review by war-in — APPROVED** [review]
